src/MagnumPlugins/WavAudioImporter/WavImporter.cpp

```diff
@@ -43,74 +43,161 @@ bool WavImporter::doIsOpened() const { return _data; }
         return;
     }
 
-    /* Verify more things */
-    if(header.subChunk1Size != 16 ||
-       header.subChunk2Size + 44 != data.size() ||
-       header.blockAlign != header.numChannels*header.bitsPerSample/8 ||
-       header.byteRate != header.sampleRate*header.blockAlign) {
+    /* Format sanity checks */
+    if(formatChunk.blockAlign != formatChunk.numChannels * formatChunk.bitsPerSample / 8 ||
+       formatChunk.byteRate != formatChunk.sampleRate * formatChunk.blockAlign) {
         Error() << "Audio::WavImporter::openData(): the file is corrupted";
         return;
     }
 
-    /* Decide about format */
-    if(header.numChannels == 1 && header.bitsPerSample == 8)
-        _format = Buffer::Format::Mono8;
-    else if(header.numChannels == 1 && header.bitsPerSample == 16)
-        _format = Buffer::Format::Mono16;
-    else if(header.numChannels == 2 && header.bitsPerSample == 8)
-        _format = Buffer::Format::Stereo8;
-    else if(header.numChannels == 2 && header.bitsPerSample == 16)
-        _format = Buffer::Format::Stereo16;
-    else {
-        Error() << "Audio::WavImporter::openData(): unsupported channel count"
-                << header.numChannels << "with" << header.bitsPerSample
-                << "bits per sample";
+    const RiffChunk* dataChunk = nullptr;
+    UnsignedInt dataChunkSize = 0;
+
+    const UnsignedInt headerSize = sizeof(WavHeaderChunk) + sizeof(RiffChunk) + formatChunk.chunk.chunkSize;
+    UnsignedInt offset = 0;
+
+    /* Skip any chunks that aren't the data chunk */
+    while(headerSize + offset <= header.chunk.chunkSize) {
+        const RiffChunk* currChunk = reinterpret_cast<const RiffChunk*>(data.begin() + headerSize + offset);
+        offset += Utility::Endianness::littleEndian(currChunk->chunkSize) + sizeof(RiffChunk);
+
+        if(std::strncmp(currChunk->chunkId, "data", 4) == 0) {
+            dataChunk = currChunk;
+            dataChunkSize = Utility::Endianness::littleEndian(currChunk->chunkSize);
+
+            break;
+        }
+    }
+
+    /* Make sure we actually got a data chunk */
+    if(dataChunk == nullptr) {
+        Error() << "Audio::WavImporter::openData(): the file contains no data chunk";
+        return;
+    }
+
+    /* Size sanity checks */
+    if(headerSize + offset > data.size()) {
+        Error() << "Audio::WavImporter::openData(): file size doesn't match computed size";
         return;
     }
 
     /* Save frequency */
-    _frequency = header.sampleRate;
+    _frequency = formatChunk.sampleRate;
 
     /** @todo Convert the data from little endian too */
     CORRADE_INTERNAL_ASSERT(!Utility::Endianness::isBigEndian());
 
     /* Copy the data */
-    _data = Containers::Array<char>(header.subChunk2Size);
-    std::copy(data.begin()+sizeof(WavHeader), data.end(), _data.begin());
+    const char* dataChunkPtr = reinterpret_cast<const char*>(++dataChunk);
```

Comment:
I would prefer `dataChunk + 1` instead of `++dataChunk`, as it doesn't change the original value.